Blaze upgrade

moxcodes opened the pull request:
## Proposed changes

This was more of a hassle than I anticipated going into it. Blaze changed around a bunch of their template expression type aliases to use their `Bind1st` and `Bind2nd` structs. Also, they changed their matrix class to be padded by default, which broke all of our lapack calls.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

moxcodes committed:
Introduce compatibility with Blaze 3.5-3.7

- Drop compatibility with Blaze 3.4 and earlier
- Add note about blaze 3.7 matrix padding
- Update container version to Blaze 3.7

nilsdeppe commented on src/DataStructures/ComplexDataVector.hpp:
```diff
@@ -63,28 +63,6 @@ class ComplexDataVector
 namespace blaze {
 VECTOR_BLAZE_TRAIT_SPECIALIZE_ARITHMETIC_TRAITS(ComplexDataVector);
 VECTOR_BLAZE_TRAIT_SPECIALIZE_ALL_MAP_TRAITS(ComplexDataVector);
-#if ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION <= 3))
```
I haven't had a chance to look through this in detail, but can you bump the required version in `cmake/SetupBlaze.cmake` and in our installation instructions?

moxcodes commented:
Posted a fixup with the rest of the version bump bookkeeping.
Cheers!

nilsdeppe commented on cmake/SetupBlaze.cmake:
```diff
@@ -1,7 +1,7 @@
 # See LICENSE.txt for details.
 
-find_package(Blaze 3.2 REQUIRED)
+find_package(Blaze 3.5 REQUIRED)
```
I'm happy for you to squash this

moxcodes commented:
Squashed. I also added one more commit with some explanation and instruction about the Blaze trouble to `DenseMatrix` so that there is at least somewhere to find the information the next time someone finds a mysteriously broken LAPACK call.
Cheers!

kidder commented on src/Utilities/PointerVector.hpp:
```diff
@@ -28,12 +29,21 @@
 // respective versions.
 #if ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION == 2))
 const bool blaze_unaligned = blaze::unaligned;
+using AF_t = bool;
```
AF_t -> AlignmentFlag_t

kidder commented on src/DataStructures/ComplexModalVector.hpp:
```diff
@@ -136,7 +136,7 @@ struct BinaryMapTrait<ComplexModalVector, ComplexModalVector, Operator> {
   using Type = ComplexModalVector;
 };
-#else
+#elif ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION == 4))
```
why is this just 4 and not <6?

kidder commented on src/DataStructures/DenseMatrix.hpp:
```diff
@@ -16,6 +16,7 @@
 #include <blaze/math/DynamicMatrix.h>
 #include <blaze/system/Version.h>
 
+
```
huh?

kidder commented on src/Utilities/PointerVector.hpp:
```diff
@@ -29,19 +29,33 @@
 const bool blaze_unaligned = blaze::unaligned != 0;
+const bool blaze_unpadded = blaze::unpadded != 0;
 using AF_t = bool;
+using PF_t = bool;
```
PF_t -> PaddingFlag_t

kidder commented on tests/Unit/RunSingleTest/CMakeLists.txt:
```diff
@@ -32,7 +32,7 @@ add_spectre_executable(
   # The following lines list the source files with tests in them to be
   # compiled into the executable.
-  ../DataStructures/Test_DataVector.cpp
+  ../DataStructures/Test_DenseMatrix.cpp
```
don't change this

kidder commented on tests/Unit/TestHelpers.hpp:
```diff
@@ -114,9 +114,9 @@ template <typename T, Requires<tt::has_equivalence<T>::value> = nullptr,
   static_assert(std::is_rvalue_reference<decltype(a)>::value,
                 "Must move into test_move_semantics");
-  static_assert(std::is_nothrow_move_assignable<T>::value,
+  static_assert(std::is_move_assignable<T>::value,
```
why did you change these?

moxcodes replied:
Ah, perhaps this should have been a different commit.
The reason is that some of the types now tested with this are not nothrow assignable and this assert fails. Do you prefer creating another utility to support those types?

nilsdeppe replied:
The types only copy pointers, why should that ever throw?

moxcodes replied:
I agree this should work. As far as I can tell, it is entirely related to the storage order change, the `is_nothrow_move_assignable` succeeds when `blaze::defaultStorageOrder` is used, and fails now that I've defaulted it to `columnMajor`.
It also appears that this problem existed previously - if you change the default storage order to `blaze::columnMajor` in develop and try to compile with Blaze 3.2, it fails the same nothrow assert. I haven't been able to track down precisely _why_ it behaves in this way, but I can only imagine that it is a subtle Blaze inheritance defect.
How would you like me to proceed for this?

kidder reviewed (changes requested):
joint review with @nilsdeppe

moxcodes commented:
The flag changes necessitated putting the fixups interleaved with the existing commits. The other alterations needed to be squashed in directly. The only unaddressed comment should be the one regarding `nothrow_move_assignable`.
Cheers!

moxcodes commented:
rebased following test changes (though this one will fail all the CI regardless due to the library version changes...)

moxcodes commented:
Pushed one more fixup regarding the unfortunate Blaze throwing behavior.
Cheers!

nilsdeppe reviewed (commented):
LGTM, once @kidder is happy please go ahead and squash, then we can coordinate merging this

kidder commented:
@moxcodes you can squash

moxcodes commented:
fixups squashed in.
Cheers!

moxcodes commented:
Squashed down to one commit.
Cheers!

moxcodes commented:
Squashed in a warning suppression in `NumericalAlgorithms/LinearSolver/Lapack.cpp`, one of the gcc's warned about a redundant declaration with a declaration in blaze. The only change was the three `#pragma` lines around the lapack declaration.
Cheers!

moxcodes commented:
Squashed in a fix to `DataStructures/Python/Matrix.cpp` because apparently the padding thing also broke python bindings.
Cheers!
